api: Create experiment abstraction for closed beta features

victorges opened the pull request:
<!-------------------------------------------------------------------------
 | Thanks for send a pull request! 🎉
 | First, please make sure you familiar with the contribution guidelines
 | https://github.com/livepeer/livepeer.studio/blob/master/CONTRIBUTING.md
 -------------------------------------------------------------------------->

**What does this pull request do? Explain your changes. (required)**

This is to create a new abstraction called "experiment" to allow us to iterate on
private experiments with specific customers without risking having a feature widely
available to all the users.

This was originally designed here: https://www.notion.so/livepeer/Feature-Experimentation-Playbook-f80717040b254c1ca3591169bcf6611f

**Specific updates (required)**
 - Create new experiment schema/table
 - Create support for GIN indexes to use in that table
 - Create new experiment API CRUD
 - Create specific endpoints for the more useful experiment actions (checking, adding/removing users etc)

**How did you test each of these updates (required)**
`yarn test`, also putting on staging to test the internal api

**Does this pull request close any open issues?**
Implements #1617 

**Checklist**

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

victorges committed:
api: Create experiment API to check a user

victorges committed:
api: Register experiment handler

victorges committed:
WIP gatekeep

victorges committed:
api: Create experiment abstraction with GIN index

For querying all experiments from an user quickly!

victorges committed:
api: Implement experiments via an API obj instead

victorges committed:
api: Implement API to change experiment audience

victorges committed:
api: Create simple experiment CRUD API

victorges committed:
api: Fix the schema

victorges committed:
api: Add user ID to experiment object

victorges committed:
api: Create createdAt and updatedAt fields

victorges committed:
api: Create generic listing API to replace by-user endpoint

victorges committed:
api: Create sub-set of experimental API endpoints

victorges committed:
api: Make experiments API experiment-gated

victorges committed:
api: Make isExperimentSubjet a static helper

To be used in experimenting code

victorges committed:
api: Make the experiment check API authorized

Always safer to start closed.

victorges committed:
api: Create some tests for experiment table index

victorges committed:
api: nit return next

gioelecerati commented on packages/api/src/controllers/access-control.ts:
```diff
@@ -12,14 +11,11 @@ import {
+app.use("/signing-key", signingKeyApp);
 
-accessControl.post(
+app.post(
```
Thank you for standardizing this

gioelecerati commented on packages/api/src/schema/schema.yaml:
```diff
@@ -2309,6 +2329,54 @@ components:
+      properties:
+        type:
+          type: string
+          enum: [experiment]
```
Is this enum needed? Are there gonna be ever other types in the experiment table?

victorges replied:
I followed the pattern from some old schemas here, but I think it's not necessary anymore! Probably only used when we had a single table for everything in some NoSQL databases

gioelecerati commented on packages/api/src/schema/schema.yaml:
```diff
@@ -2309,6 +2329,54 @@ components:
+          type: number
+          readOnly: true
+          description:
+            Timestamp (in milliseconds) at which the signing-key was created
```
Left `signing-key` into the description :D

victorges replied:
oops

gioelecerati commented on packages/api/src/controllers/experiment.ts:
```diff
@@ -0,0 +1,225 @@
+  app.use(`/api/${experiment}`, experimentSubjectsOnly(experiment), api);
+}
+
+app.get("/check/:experiment", authorizer({}), async (req, res) => {
```
Maybe we want to add some `experiment.test.js` at least for this only nonAdmin function

victorges replied:
Good point! Added some tests to it.

victorges committed:
amend to base

victorges committed:
api: Fix experiment schema

victorges committed:
api/package.json: Upgrade uuid package

victorges committed:
api: Fix uuid lib imports

victorges commented on packages/api/src/controllers/experiment.ts:
```diff
@@ -0,0 +1,215 @@
+  };
+
+for (const [experiment, api] of Object.entries(experimentApis)) {
+  app.use(`/api/${experiment}`, experimentSubjectsOnly(experiment), api);
```
I'm unhappy with the final paths of the experimental APIs. e.g.: `/api/experiment/api/lit-signing-condition/verify-lit-jwt`

I think the double `/api` looks weird but don't have many other ideas. Do you think `/x` would be too bad? i.e. `/api/experiment/x/lit-signing-condition/verify-lit-jwt`

Or otherwise I take any other ideas y'all might have!

gioelecerati replied:
Double `/api` definitely looks weird. Also `/x/` looks weird to me. What about using a number instead? Like `/1/` or `/1.0/` acting like an api manager version, so it's something that developers are used to encounter?

victorges replied:
Hmm yeah, numbers would not be too bad but it'd seem pretty artificial to me, just to have some dummy value that we never update haha

victorges replied:
Hmmm what about an explicitly dummy `-`?
`/api/experiment/-/lit-signing-condition/verify-lit-jwt`

victorges replied:
Gone with `-` :)

victorges committed:
api: Add some unrelated logs to parseFilters :see_no_evil:

I wanna make the filters actually fail the request if they
are malformed. It's very confusing to get a response that
is completely ignoring some input you provided.

victorges committed:
api: Fix schema based on PR comments

victorges committed:
api: Use - as the experimental api separator

victorges committed:
api: Add tests to experiment API
